src/app/invest/invest.component.ts

```diff
@@ -19,31 +21,71 @@ import {RouterService} from '../shared/services/router.service'
 export class InvestComponent {
   investState = InvestState
 
-  campaign$: Observable<WithStatus<CampaignWithInfo>>
-  balance$ = withStatus(
-    this.stablecoinService.balance$.pipe(
-      shareReplay(1),
-    ),
+  campaign$: Observable<CampaignWithInfo>
+  campaignWithStatus$: Observable<WithStatus<CampaignWithInfo>>
+
+  balance$ = this.stablecoinService.balance$.pipe(
+    shareReplay(1),
   )
 
+  alreadyInvested$: Observable<number>
+  preInvestData$: Observable<PreInvestData>
+
   investStateSub = new BehaviorSubject<InvestState>(InvestState.Editing)
   investState$ = this.investStateSub.asObservable()
 
   investmentForm: FormGroup
 
   constructor(private fb: FormBuilder,
               private campaignService: CampaignService,
+              private sessionQuery: SessionQuery,
               private stablecoinService: StablecoinService,
               private dialogService: DialogService,
               private router: RouterService,
               private route: ActivatedRoute) {
     const campaignID = this.route.snapshot.params.id
 
-    this.campaign$ = withStatus(
-      this.campaignService.getAddressByName(campaignID).pipe(
-        switchMap(address => this.campaignService.getCampaignWithInfo(address)),
-        shareReplay(1),
-      ),
+    this.campaign$ = this.campaignService.getAddressByName(campaignID).pipe(
+      switchMap(address => this.campaignService.getCampaignWithInfo(address)),
+      shareReplay(1),
+    )
+
+    this.campaignWithStatus$ = withStatus(this.campaign$)
+
+    this.alreadyInvested$ = this.campaign$.pipe(
+      switchMap(campaign =>
+        this.campaignService.alreadyInvested(campaign.contractAddress)),
+      shareReplay(1),
+    )
+
+    this.preInvestData$ = combineLatest([
+      this.campaign$,
+      this.balance$,
+      this.alreadyInvested$,
+    ]).pipe(take(1),
+      map(([campaign, balance, alreadyInvested]) => {
+        const campaignMin = Number(utils.formatEther(campaign.minInvestment))
+        const campaignMax = Number(utils.formatEther(campaign.maxInvestment))
+
+        const campaignTokensBalance = Number(utils.formatEther(campaign.totalTokensBalance))
+        const campaignTokensSold = Number(utils.formatEther(campaign.totalTokensSold))
+        const campaignTokenPrice = TokenPrice.parse(campaign.tokenPrice.toNumber())
+        const campaignTokensAvailable = campaignTokensBalance - campaignTokensSold
+        const campaignAmountAvailable = campaignTokensAvailable * campaignTokenPrice
+
+        const walletBalance = Number(utils.formatEther(balance))
+
+        let min = alreadyInvested > 0 ? 0 : campaignMin
+        let max = campaignAmountAvailable < campaignMax ? campaignAmountAvailable : campaignMax
+
+        // ensure floor on 2 decimal points
+        max = Math.round(Math.floor(max * 100)) / 100
```

Comment:
Why use `round` after `floor`? Looks to me that `floor` produces a round number: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/floor

Reply:
Sure!
```
Math.floor(max * 100) / 100
```